xaitk_saliency/interfaces/gen_object_detector_blackbox_sal.py

```diff
@@ -0,0 +1,175 @@
+            reference detection. This is a single matrix of shape
+            `[nDets x H x W]` where `H` and `W` are the height and width
+            respectively of the input reference image.
+        """
+
+        if ref_image.ndim not in (2, 3):
+            raise ValueError(f"Input image matrix has an unexpected number of dimensions: {ref_image.ndim}")
+
+        if bboxes.shape[1] != 4:
+            raise ValueError(f"Input bounding boxes matrix has width of {bboxes.shape[1]}, should have width of 4")
+
+        # Check that reference detection inputs have matching shapes
+        if objectness is None:
+            if len(bboxes) != len(scores):
+                raise ValueError(
+                    f"Number of input bounding boxes and scores do not match: "
+                    f"(bboxes) {len(bboxes)} != {len(scores)} (scores)"
+                )
+        else:
+            if len(bboxes) != len(scores) or len(bboxes) != len(objectness):
+                raise ValueError(
+                    f"Number of input bounding boxes, scores, and objectness "
+                    f"scores do not match: (bboxes) {len(bboxes)} != "
+                    f"{len(scores)} (scores) and/or (bboxes) {len(bboxes)} != "
+                    f"{len(objectness)} (objectness)"
+                )
+
+        output = self._generate(
+            ref_image,
+            bboxes,
+            scores,
+            blackbox,
+            objectness,
+        )
+
+        # Check that the saliency heatmaps' shape matches the reference image.
+        if output.shape[1:] != ref_image.shape[:2]:
+            raise ShapeMismatchError(
+                f"Output saliency heatmaps did not have matching height and "
+                f"width shape components: "
+                f"(ref) {ref_image.shape[:2]} != {output.shape[1:]} (output)"
+            )
+        return output
+
+    def __call__(
+        self,
+        ref_image: np.ndarray,
+        bboxes: np.ndarray,
+        scores: np.ndarray,
+        blackbox: DetectImageObjects,
+        objectness: Optional[np.ndarray] = None,
+    ) -> np.ndarray:
+        """
+        Alias to the :meth:`generate` method.
+        See :meth:`generate` for more details.
+        """
+        return self.generate(
+            ref_image,
+            bboxes,
+            scores,
+            blackbox,
+            objectness,
+        )
+
+    @abc.abstractmethod
+    def _generate(
+        self,
+        ref_image: np.ndarray,
+        bboxes: np.ndarray,
+        scores: np.ndarray,
+        blackbox: DetectImageObjects,
+        objectness: Optional[np.ndarray] = None,
+    ) -> np.ndarray:
+        """
+        Internal method for implementing the generation logic.
+        This is invoked by the above `generate` method as a template method.
+
+        The doc-string for the `generate` method also applies here aside from
+        the `ShapeMismatchError` which is specific to `generate`.
```

Comment:
We could also add that certain guarantees are being levied on the input parameters given the ValueError checks in `generate`.